src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/impl/DynamicIncludeFlatBuilder.java

```diff
@@ -23,28 +44,94 @@ public void evaluate(TemplateOutputWriter writer, TemplateBuilderContext context
             evaluate = evaluateExpressions(cql, context);
         }
 
-        // updating the key and the value since we do not want to display includeFlat keyword in the
-        // result
-        if (evaluate instanceof ObjectNode) {
-            if (((ObjectNode) evaluate).fields().hasNext()) {
-                Map.Entry<String, JsonNode> includeFlatField =
-                        ((ObjectNode) evaluate).fields().next();
-                setKey(includeFlatField.getKey());
-                evaluate = includeFlatField.getValue();
-            }
+        ObjectNode finalNode = null;
+        if (!(evaluate instanceof ObjectNode) && evaluate != null) {
+            String message = "Cannot include flat a value different from a JSON object";
+            LOGGER.severe(() -> message);
+            throw new UnsupportedOperationException(message);
+        } else if (evaluate != null) {
+            JSONMerger merger = new JSONMerger();
+            finalNode = merger.mergeTrees(includingNode, (ObjectNode) evaluate);
+        }
+
+        if (finalNode != null) doIncludeFlat(finalNode, context, writer);
+        else
+            // write the node as it is
+            iterateAndWrite((ObjectNode) includingNode, writer, context);
+    }
+
+    private void doIncludeFlat(
+            ObjectNode finalNode, TemplateBuilderContext context, TemplateOutputWriter writer)
+            throws IOException {
+        if (hasDynamic(finalNode)) {
+            LOGGER.fine(
+                    () ->
+                            "Included Json object has property interpolation or expression."
+                                    + "Going to build a nested TemplateBuilder tree.");
+            iterateAndEvaluateNestedTree(context, writer, finalNode);
+        } else {
+            LOGGER.fine(() -> "Writing the included flat Json Node.");
+            // iterate here instead of directly in writer to avoid
+            // a JSON object being started since we are writing attributes directly into the
+            // parent one.
+            iterateAndWrite(finalNode, writer, context);
+        }
+    }
+
+    private void iterateAndWrite(
+            ObjectNode objectNode, TemplateOutputWriter writer, TemplateBuilderContext context)
+            throws IOException {
+        Iterator<String> names = objectNode.fieldNames();
+        while (names != null && names.hasNext()) {
+            String name = names.next();
+            writeValue(name, writer, objectNode.findValue(name), context);
         }
+    }
 
-        if (evaluate instanceof JsonNode && hasDynamic((JsonNode) evaluate))
-            writeFromNestedTree(context, writer, (JsonNode) evaluate);
-        else writeValue(writer, evaluate, context);
+    void iterateAndEvaluateNestedTree(
+            TemplateBuilderContext context, TemplateOutputWriter writer, ObjectNode node)
+            throws IOException {
+        TemplateReaderConfiguration configuration =
```

Comment:
For the life of me I don't understand what's going on in this method! :-D
Maybe some comments would help.